File: Assets/Scripts/DouyinSequenceController.lua
---@var Registry               :DouyinScript
---@var DelayAfterPendulum     :float = 0
---@var DelayAfterExtend       :float = 0
---@var DelayAfterBallRotate   :float = 0
---@var DelayAfterCenterSpin   :float = 0
---@var DelayAfterBallReset    :float = 0
---@var DelayAfterRetract      :float = 0
---@var AutoPlay               :bool = true
---@var LoopSequence           :bool = true
---@var EnableDebugLog         :bool = false
---@end

local STATE_IDLE = 0
local STATE_PENDULUM = 1
local STATE_EXTEND = 2
local STATE_BALL_ROTATE = 3
local STATE_CENTER_SPIN = 4
local STATE_BALL_RESET = 5
local STATE_RETRACT = 6
local STATE_WAIT = 7

local currentState = STATE_IDLE
local isPlaying = false
local waitTimer = 0
local waitNextState = STATE_IDLE

local function Log(message)
    if EnableDebugLog then
        print("[DouyinSequenceController] " .. tostring(message))
    end
end

local function ClampNonNegative(value, defaultValue)
    if value == nil then
        return defaultValue
    end
    if value < 0 then
        return 0
    end
    return value
end

local function HasRegistry()
    return Registry ~= nil and Registry.script ~= nil
end

local function EnterState(state)
    currentState = state
    Log("EnterState=" .. tostring(state))
end

local function BeginState(state)
    if not HasRegistry() then
        EnterState(STATE_IDLE)
        isPlaying = false
        return
    end

    if state == STATE_PENDULUM then
        Registry.script.BeginPendulums()
    elseif state == STATE_EXTEND then
        Registry.script.BeginExtendForward()
    elseif state == STATE_BALL_ROTATE then
        Registry.script.BeginBallForward()
    elseif state == STATE_CENTER_SPIN then
        Registry.script.BeginOrbit()
        Registry.script.BeginOrbitFx(Registry.script.GetOrbitDuration())
    elseif state == STATE_BALL_RESET then
        Registry.script.BeginBallBackward()
    elseif state == STATE_RETRACT then
        Registry.script.BeginExtendBackward()
    end

    EnterState(state)
end

local function EnterWaitState(duration, nextState)
    local waitDuration = ClampNonNegative(duration, 0)
    if waitDuration <= 0 then
        BeginState(nextState)
        return
    end

    waitTimer = waitDuration
    waitNextState = nextState
    EnterState(STATE_WAIT)
end

function Awake()
    if AutoPlay == nil then
        AutoPlay = true
    end
    if LoopSequence == nil then
        LoopSequence = true
    end
    if EnableDebugLog == nil then
        EnableDebugLog = false
    end

    if HasRegistry() then
        Registry.script.Validate()
        Registry.script.CollectAll()
        Registry.script.ResetAllToBase()
        Registry.script.SetAllVertexLightIntensityScale(Registry.script.GetCurrentNightBlend01())
    else
        Log("Registry is nil or Registry.script is nil")
    end

    isPlaying = AutoPlay and HasRegistry()
    if isPlaying then
        BeginState(STATE_PENDULUM)
    else
        EnterState(STATE_IDLE)
    end
end

function Update()
    if not isPlaying or not HasRegistry() then
        return
    end

    local deltaTime = CS.UnityEngine.Time.deltaTime
    Registry.script.TickVertexLights(deltaTime)

    local gearDriveFactor = nil
    if currentState == STATE_CENTER_SPIN then
        gearDriveFactor = Registry.script.GetOrbitSpeedFactor()
    end
    Registry.script.TickGears(deltaTime, gearDriveFactor)

    if currentState == STATE_WAIT then
        waitTimer = waitTimer - deltaTime
        if waitTimer <= 0 then
            waitTimer = 0
            BeginState(waitNextState)
        end
        return
    end

    if currentState == STATE_PENDULUM then
        Registry.script.TickPendulums(deltaTime)
        if Registry.script.ArePendulumsComplete() then
            EnterWaitState(DelayAfterPendulum, STATE_EXTEND)
        end
        return
    end

    if currentState == STATE_EXTEND then
        Registry.script.TickExtenders(deltaTime)
        if Registry.script.AreExtendersComplete() then
            EnterWaitState(DelayAfterExtend, STATE_BALL_ROTATE)
        end
        return
    end

    if currentState == STATE_BALL_ROTATE then
        Registry.script.TickBalls(deltaTime)
        if Registry.script.AreBallsComplete() then
            EnterWaitState(DelayAfterBallRotate, STATE_CENTER_SPIN)
        end
        return
    end

    if currentState == STATE_CENTER_SPIN then
        Registry.script.TickOrbit(deltaTime)
        local orbitProgress = Registry.script.GetOrbitProgress01()
        Registry.script.TickOrbitFx(deltaTime, orbitProgress)
        Registry.script.SetAllVertexLightIntensityScale(Registry.script.GetOrbitNightBlend01(orbitProgress))
        if Registry.script.IsOrbitComplete() then
            Registry.script.CompleteOrbitFx()
            EnterWaitState(DelayAfterCenterSpin, STATE_BALL_RESET)
        end
        return
    end

    if currentState == STATE_BALL_RESET then
        Registry.script.TickBalls(deltaTime)
        if Registry.script.AreBallsComplete() then
            EnterWaitState(DelayAfterBallReset, STATE_RETRACT)
        end
        return
    end

    if currentState == STATE_RETRACT then
        Registry.script.TickExtenders(deltaTime)
        if Registry.script.AreExtendersComplete() then
            if LoopSequence then
                EnterWaitState(DelayAfterRetract, STATE_PENDULUM)
            else
                Registry.script.ResetAllToBase()
                Registry.script.SetAllVertexLightIntensityScale(Registry.script.GetCurrentNightBlend01())
                EnterState(STATE_IDLE)
                isPlaying = false
            end
        end
    end
end

function PlaySequence()
    if not HasRegistry() then
        Log("PlaySequence skipped because Registry is unavailable")
        return
    end

    Registry.script.CollectAll()
    if currentState == STATE_IDLE then
        BeginState(STATE_PENDULUM)
    end
    isPlaying = true
end

function PauseSequence()
    isPlaying = false
end

function ResumeSequence()
    if HasRegistry() then
        isPlaying = true
    end
end

function RestartSequence()
    if not HasRegistry() then
        return
    end

    Registry.script.CollectAll()
    Registry.script.ResetAllToBase()
    Registry.script.SetAllVertexLightIntensityScale(Registry.script.GetCurrentNightBlend01())
    waitTimer = 0
    waitNextState = STATE_IDLE
    isPlaying = true
    BeginState(STATE_PENDULUM)
end

function StopSequence()
    isPlaying = false
    waitTimer = 0
    waitNextState = STATE_IDLE

    if HasRegistry() then
        Registry.script.ResetAllToBase()
        Registry.script.SetAllVertexLightIntensityScale(Registry.script.GetCurrentNightBlend01())
    end

    EnterState(STATE_IDLE)
end

function GetCurrentState()
    return currentState
end

File: Assets/Scripts/DouyinSequenceController.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using UnityEngine;

public sealed class DouyinSequenceController : MonoBehaviour
{
    private enum SequenceState
    {
        Idle,
        Pendulum,
        ExtendForward,
        BallRotate,
        CenterSpin,
        BallReset,
        Retract,
        Wait
    }

    public DouyinSequenceRegistry Registry;
    public float DelayAfterPendulum;
    public float DelayAfterExtend;
    public float DelayAfterBallRotate;
    public float DelayAfterCenterSpin;
    public float DelayAfterBallReset;
    public float DelayAfterRetract;
    public bool AutoPlay = true;
    public bool LoopSequence = true;
    public bool EnableDebugLog;

    private SequenceState currentState = SequenceState.Idle;
    private bool isPlaying;
    private bool pendingAutoPlay;
    private float waitTimer;
    private SequenceState waitNextState = SequenceState.Idle;

    private void Awake()
    {
        if (Registry != null)
        {
            Registry.Validate();
            Registry.CollectAll();
            Registry.ResetAllToBase();
            Registry.SetAllVertexLightIntensityScale(Registry.GetCurrentNightBlend01());
        }
        else
        {
            Log("Registry is nil");
        }

        pendingAutoPlay = AutoPlay && Registry != null;
        isPlaying = false;
        EnterState(SequenceState.Idle);
    }

    private void Update()
    {
        if (pendingAutoPlay)
        {
            pendingAutoPlay = false;
            if (Registry != null)
            {
                isPlaying = true;
                BeginState(SequenceState.Pendulum);
            }
        }

        if (!isPlaying || Registry == null)
        {
            return;
        }

        float deltaTime = Time.deltaTime;
        Registry.TickVertexLights(deltaTime);
        Registry.TickSequenceAudio(deltaTime);

        float? gearDriveFactor = currentState == SequenceState.CenterSpin ? Registry.GetOrbitSpeedFactor() : null;
        Registry.TickGears(deltaTime, gearDriveFactor);

        if (currentState == SequenceState.Wait)
        {
            waitTimer -= deltaTime;
            if (waitTimer <= 0f)
            {
                waitTimer = 0f;
                BeginState(waitNextState);
            }
            return;
        }

        if (currentState == SequenceState.Pendulum)
        {
            Registry.TickPendulums(deltaTime);
            if (Registry.ArePendulumsComplete())
            {
                EnterWaitState(DelayAfterPendulum, SequenceState.ExtendForward);
            }
            return;
        }

        if (currentState == SequenceState.ExtendForward)
        {
            Registry.TickExtenders(deltaTime);
            if (Registry.AreExtendersComplete())
            {
                EnterWaitState(DelayAfterExtend, SequenceState.BallRotate);
            }
            return;
        }

        if (currentState == SequenceState.BallRotate)
        {
            Registry.TickBalls(deltaTime);
            if (Registry.AreBallsComplete())
            {
                EnterWaitState(DelayAfterBallRotate, SequenceState.CenterSpin);
            }
            return;
        }

        if (currentState == SequenceState.CenterSpin)
        {
            Registry.TickOrbit(deltaTime);
            float orbitProgress = Registry.GetOrbitProgress01();
            Registry.TickOrbitFx(deltaTime, orbitProgress);
            Registry.SetAllVertexLightIntensityScale(Registry.GetOrbitNightBlend01(orbitProgress));
            if (Registry.IsOrbitComplete())
            {
                Registry.CompleteOrbitFx();
                EnterWaitState(DelayAfterCenterSpin, SequenceState.BallReset);
            }
            return;
        }

        if (currentState == SequenceState.BallReset)
        {
            Registry.TickBalls(deltaTime);
            if (Registry.AreBallsComplete())
            {
                EnterWaitState(DelayAfterBallReset, SequenceState.Retract);
            }
            return;
        }

        if (currentState == SequenceState.Retract)
        {
            Registry.TickExtenders(deltaTime);
            if (!Registry.AreExtendersComplete())
            {
                return;
            }

            if (LoopSequence)
            {
                EnterWaitState(DelayAfterRetract, SequenceState.Pendulum);
            }
            else
            {
                Registry.ResetAllToBase();
                Registry.SetAllVertexLightIntensityScale(Registry.GetCurrentNightBlend01());
                EnterState(SequenceState.Idle);
                isPlaying = false;
            }
        }
    }

    public void PlaySequence()
    {
        if (Registry == null)
        {
            Log("PlaySequence skipped because Registry is unavailable");
            return;
        }

        Registry.CollectAll();
        if (currentState == SequenceState.Idle)
        {
            pendingAutoPlay = false;
            BeginState(SequenceState.Pendulum);
        }
        isPlaying = true;
    }

    public void PauseSequence()
    {
        isPlaying = false;
        Registry?.PauseSequenceAudio();
    }

    public void ResumeSequence()
    {
        if (Registry == null)
        {
            return;
        }

        Registry.ResumeSequenceAudio();
        isPlaying = true;
    }

    public void RestartSequence()
    {
        if (Registry == null)
        {
            return;
        }

        Registry.CollectAll();
        Registry.ResetAllToBase();
        Registry.SetAllVertexLightIntensityScale(Registry.GetCurrentNightBlend01());
        pendingAutoPlay = false;
        waitTimer = 0f;
        waitNextState = SequenceState.Idle;
        isPlaying = true;
        BeginState(SequenceState.Pendulum);
    }

    public void StopSequence()
    {
        isPlaying = false;
        pendingAutoPlay = false;
        waitTimer = 0f;
        waitNextState = SequenceState.Idle;

        if (Registry != null)
        {
            Registry.ResetAllToBase();
            Registry.SetAllVertexLightIntensityScale(Registry.GetCurrentNightBlend01());
            Registry.StopSequenceAudio();
        }

        EnterState(SequenceState.Idle);
    }

    public int GetCurrentState()
    {
        return (int)currentState;
    }

    private void BeginState(SequenceState state)
    {
        if (Registry == null)
        {
            EnterState(SequenceState.Idle);
            isPlaying = false;
            return;
        }

        if (state == SequenceState.Pendulum)
        {
            Registry.BeginPendulums();
        }
        else if (state == SequenceState.ExtendForward)
        {
            Registry.BeginExtendForward();
        }
        else if (state == SequenceState.BallRotate)
        {
            Registry.BeginBallForward();
        }
        else if (state == SequenceState.CenterSpin)
        {
            Registry.BeginOrbit();
            Registry.BeginOrbitFx(Registry.GetOrbitDuration());
        }
        else if (state == SequenceState.BallReset)
        {
            Registry.BeginBallBackward();
        }
        else if (state == SequenceState.Retract)
        {
            Registry.BeginExtendBackward();
        }

        BeginAudioState(state);
        EnterState(state);
    }

    private void EnterWaitState(float duration, SequenceState nextState)
    {
        float waitDuration = Mathf.Max(0f, duration);
        if (waitDuration <= 0f)
        {
            BeginState(nextState);
            return;
        }

        waitTimer = waitDuration;
        waitNextState = nextState;
        BeginAudioState(SequenceState.Wait);
        EnterState(SequenceState.Wait);
    }

    private void BeginAudioState(SequenceState state)
    {
        Registry?.BeginSequenceAudio(GetStateName(state));
    }

    private void EnterState(SequenceState state)
    {
        currentState = state;
        Log("EnterState=" + state);
    }

    private static string GetStateName(SequenceState state)
    {
        switch (state)
        {
            case SequenceState.Pendulum:
                return "Pendulum";
            case SequenceState.ExtendForward:
                return "ExtendForward";
            case SequenceState.BallRotate:
                return "BallRotate";
            case SequenceState.CenterSpin:
                return "CenterSpin";
            case SequenceState.BallReset:
                return "BallReset";
            case SequenceState.Retract:
                return "Retract";
            case SequenceState.Wait:
                return "Wait";
            default:
                return "Idle";
        }
    }

    private void Log(string message)
    {
        if (EnableDebugLog)
        {
            Debug.Log("[DouyinSequenceController] " + message, this);
        }
    }
}

